dev/devicelab/lib/tasks/keyboard_hot_restart_test.dart

```diff
@@ -0,0 +1,149 @@
+      final Device device = await devices.workingDevice;
+      await device.unlock();
+      deviceIdOverride = device.deviceId;
+    }
+
+    await inDirectory<void>(appDir, () async {
+      try {
+        section('Create app');
+        await createAppProject();
+
+        // Ensure the app forces the keyboard to be visible.
+        final String newContents = oldContents.replaceFirst(forceKeyboardOff, forceKeyboardOn);
+        mainFile.writeAsStringSync(newContents);
+
+        section('Launch app and wait for keyboard to be visible');
+
+        TestState state = TestState.waitUntilKeyboardOpen;
+
+        await runApp(
+          options: <String>['-d', deviceIdOverride!],
+          onLine: (String line, Process process) {
+            if (state == TestState.waitUntilKeyboardOpen) {
+              if (!lineIndicatesKeyboardIsOpen(line)) {
+                return;
+              }
+
+              section('Update the app to no longer force the keyboard to be visible');
+              final String newContents = oldContents.replaceFirst(
+                forceKeyboardOn,
+                forceKeyboardOff,
+              );
+              mainFile.writeAsStringSync(newContents);
+
+              section('Hot reload app');
+              process.stdin.writeln('R');
+
+              section('Wait until the keyboard is no longer visible');
+              state = TestState.waitUntilKeyboardClosed;
+            } else if (state == TestState.waitUntilKeyboardClosed) {
+              if (!lineIndicatesKeyboardIsClosed(line)) {
+                return;
+              }
+
+              process.stdin.writeln('q');
+            }
+          },
+        );
+      } finally {
+        mainFile.writeAsStringSync(oldContents);
+      }
+    });
+
+    return TaskResult.success(null);
+  };
+}
+
+enum TestState { waitUntilKeyboardOpen, waitUntilKeyboardClosed }
+
+Future<void> createAppProject() async {
+  await exec(path.join(flutterDirectory.path, 'bin', 'flutter'), <String>[
+    'create',
+    '--platforms=android,ios',
+    '.',
+  ]);
+}
+
+Future<void> runApp({
+  required List<String> options,
+  required void Function(String, Process) onLine,
+}) async {
+  final Process process = await startFlutter('run', options: options);
+
+  final Completer<void> stdoutDone = Completer<void>();
+  final Completer<void> stderrDone = Completer<void>();
+
+  void onStdout(String line) {
+    onLine(line, process);
+    print('stdout: $line');
+  }
+
+  process.stdout
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .listen(onStdout, onDone: stdoutDone.complete);
+
+  process.stderr
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .listen((String line) => print('stderr: $line'), onDone: stderrDone.complete);
+
+  await Future.wait<void>(<Future<void>>[stdoutDone.future, stderrDone.future]);
+  await process.exitCode;
+}
+
+bool lineIndicatesKeyboardIsOpen(String line) {
```

Comment:
is this checking which line of code is running? or is it checking the log? or a label on the screen?

Reply:
Console output I think?

Reply:
Yup this is console output. FYI, I removed these helpers, see: https://github.com/flutter/flutter/pull/167013#discussion_r2050904793